mava/networks/sable_network.py

```diff
@@ -0,0 +1,547 @@
+from mava.networks.torsos import SwiGLU
+from mava.networks.utils.sable.discrete_trainer_executor import *  # noqa
+from mava.systems.sable.types import HiddenStates, SableNetworkConfig
+from mava.types import Observation
+
+
+class EncodeBlock(nn.Module):
+    """Sable encoder block."""
+
+    net_config: SableNetworkConfig
+    memory_config: DictConfig
+    n_agents: int
+
+    def setup(self) -> None:
+        # Initialize the RMSNorm layer normalization
+        self.ln1 = nn.RMSNorm()
+        self.ln2 = nn.RMSNorm()
+
+        # Initialize the MultiScaleRetention
+        self.retn = MultiScaleRetention(
+            embed_dim=self.net_config.embed_dim,
+            n_head=self.net_config.n_head,
+            n_agents=self.n_agents,
+            full_self_retention=True,  # Full retention for the encoder
+            memory_config=self.memory_config,
+            decay_scaling_factor=self.memory_config.decay_scaling_factor,
+        )
+
+        # Initialize SwiGLU feedforward network
+        self.ffn = SwiGLU(self.net_config.embed_dim, self.net_config.embed_dim)
+
+    def __call__(
+        self, x: chex.Array, hstate: chex.Array, dones: chex.Array, timestep_id: chex.Array
+    ) -> chex.Array:
+        """Applies Chunkwise MultiScaleRetention."""
+        ret, updated_hstate = self.retn(
+            key=x, query=x, value=x, hstate=hstate, dones=dones, timestep_id=timestep_id
+        )
+        x = self.ln1(x + ret)
+        output = self.ln2(x + self.ffn(x))
+        return output, updated_hstate
+
+    def recurrent(self, x: chex.Array, hstate: chex.Array, timestep_id: chex.Array) -> chex.Array:
+        """Applies Recurrent MultiScaleRetention."""
+        ret, updated_hstate = self.retn.recurrent(
+            key_n=x, query_n=x, value_n=x, hstate=hstate, timestep_id=timestep_id
+        )
+        x = self.ln1(x + ret)
+        output = self.ln2(x + self.ffn(x))
+        return output, updated_hstate
+
+
+class Encoder(nn.Module):
+    """Multi-block encoder consisting of multiple `EncoderBlock` modules."""
+
+    net_config: SableNetworkConfig
+    memory_config: DictConfig
+    n_agents: int
+
+    def setup(self) -> None:
+        # Initialize the RMSNorm layer normalization
+        self.ln = nn.RMSNorm()
+
+        # Initialize the observation encoder and value head layers
+        self.obs_encoder = nn.Sequential(
+            [
+                nn.RMSNorm(),
+                nn.Dense(
+                    self.net_config.embed_dim, kernel_init=orthogonal(jnp.sqrt(2)), use_bias=False
+                ),
+                nn.gelu,
+            ],
+        )
+        self.head = nn.Sequential(
+            [
+                nn.Dense(self.net_config.embed_dim, kernel_init=orthogonal(jnp.sqrt(2))),
+                nn.gelu,
+                nn.RMSNorm(),
+                nn.Dense(1, kernel_init=orthogonal(0.01)),
+            ],
+        )
+
+        # Initialize the encoder blocks
```

Comment:
Again tell me if you feel differently but I feel these are redundant
```suggestion
        self.ln = nn.RMSNorm()

        self.obs_encoder = nn.Sequential(
            [
                nn.RMSNorm(),
                nn.Dense(
                    self.net_config.embed_dim, kernel_init=orthogonal(jnp.sqrt(2)), use_bias=False
                ),
                nn.gelu,
            ],
        )
        self.head = nn.Sequential(
            [
                nn.Dense(self.net_config.embed_dim, kernel_init=orthogonal(jnp.sqrt(2))),
                nn.gelu,
                nn.RMSNorm(),
                nn.Dense(1, kernel_init=orthogonal(0.01)),
            ],
        )

```